core/state/statedb.go

```diff
@@ -602,3 +605,39 @@ func (s *StateDB) CommitTo(dbw trie.DatabaseWriter, deleteEmptyObjects bool) (ro
 	log.Debug("Trie cache stats after commit", "misses", trie.CacheMisses(), "unloads", trie.CacheUnloads())
 	return root, err
 }
+
+func HasDataCallback(root common.Hash, dbr hashtree.DatabaseReader) func(position, hash []byte) bool {
+	db := hashtree.NewReader(dbr, DbPrefix)
+	t, err := trie.New(root, db)
+	if err != nil {
+		panic(err)
+	}
+	return func(position, hash []byte) bool {
+		lp := len(position)
+		if lp < 33 || (lp == 33 && position[32] < 5) {
+			return t.HasData(position, hash)
+		}
+		addrHash := position[:32]
+		enc, err := t.TryGet(addrHash)
+		if len(enc) == 0 || err != nil {
+			return false
+		}
+
+		var data Account
+		if err := rlp.DecodeBytes(enc, &data); err != nil {
+			return false
+		}
+		if lp == 33 && position[32] == 5 {
+			return bytes.Equal(hash, data.CodeHash)
+		}
+		if position[32] != 6 {
```

Comment:
Many `magic` values here, please define them as some sort of consts or enum (iota?) or whatever. Hard to understand otherwise.